diag/diag.proto

```diff
@@ -0,0 +1,236 @@
+//
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// This file defines the gNOI APIs used to perform diagnostic operations on a
+// switching box.
+syntax = "proto3";
+
+package gnoi.diag;
+
+// The Diag service exports to main set of RPCs:
+// 1- BERT related RPCs: Used to perform Bit Error Rate Test (BERT)
+//    operations on a set of ports.
+// 2- BURNING related RPCs: Used to perform a vendor-provided Burnin test on the
+//    switch to ensure the box is ready to start serving traffic. Burnin tests
+//    are typically run by HwOps in the datacenter, as part of turnup or repair
+//    workflow.
+service Diag {
```

Comment:
We should probably note that these are only for stateless diagnostics activities (in other words, these activities will all end on their own). None of these should hold any state on the device. Stateful configuration should use the OpenConfig model with GNMI. I only mention this as other vendors support a stateful internal BERT and I am hoping we can expand the current OpenConfig models to fully support this. There is a test-signal boolean leaf defined at the moment though.

Reply:
Sure these are stateless operations and them failing/passing should not have a permanent artifact on the switch (unless there is something wrong HW-wise). Added a point on that to the comments. 

However I am also interested to learn more about the statefull BERT you mentioned here. What is it and how/when it is used? Any pointers?

Reply:
Here is a pointer to the leaf:
https://github.com/openconfig/public/blob/e8981a183157a5f730f3d054da63f9314f43efb4/release/models/optical-transport/openconfig-terminal-device.yang#L754

Essentially it is very similar, but the devices support permanently turning on a test signal which sends PRBS from the port. This is useful during capacity adds and sometimes troubleshooting to see if a link or portion of a link is error free. The optical vendors I have seen just do this as stateful config instead of running for specified time.